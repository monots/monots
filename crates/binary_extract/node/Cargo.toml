[package]
edition = "2021"
name = "binary_extract_node"
version = "0.0.0"
authors = ["Ifiok Jr. <ifiokotung@gmail.com>"]
publish = false

[lib]
crate-type = ["cdylib", "rlib"]


[dependencies]
anyhow = "1.0.57"
binary_extract = { path = "../" }
napi = { version = "2.5.0", default-features = false, features = ["napi4"] }
napi-derive = "2.5.0"
serde = "1.0.137"
serde_json = "1.0.81"

[build-dependencies]
napi-build = "2.0.1"
