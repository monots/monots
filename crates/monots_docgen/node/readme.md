# skribble_css_node

> `skribble_css` for node using napi.
