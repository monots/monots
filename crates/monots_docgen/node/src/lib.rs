use std::path::Path;

use napi::{bindgen_prelude::AsyncTask, Env, JsUndefined, Result, Task};
use napi_derive::napi;

pub struct Extract {
  from: String,
  to: String,
}

#[napi]
impl Task for Extract {
  type JsValue = JsUndefined;
  type Output = ();

  fn compute(&mut self) -> Result<Self::Output> {
    monots_docgen::extract(Path::new(self.from.as_str()), Path::new(self.to.as_str())).unwrap();
    Ok(())
  }

  fn resolve(&mut self, env: Env, _: ()) -> Result<Self::JsValue> {
    env.get_undefined()
  }
}

#[napi]
pub fn extract(from: String, to: String) -> AsyncTask<Extract> {
  AsyncTask::new(Extract { from, to })
}
