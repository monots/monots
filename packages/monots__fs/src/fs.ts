import type { GlobProps } from '@monots/glob';
import { getPath, glob } from '@monots/glob';
import type { IPromisesAPI } from 'memfs/lib/promises.js';
import type { DirectoryJSON } from 'memfs/lib/volume.js';
import { Volume } from 'memfs/lib/volume.js';
import * as fsPromises from 'node:fs/promises';
import path from 'node:path';

import type { AnyFunction, Method, NativeFsPromises as NativeFsPromises } from './types';
import { METHODS } from './types';

type FileSystem = 'native' | 'memory';

/**
 * Create a file system for use in monots.
 */
export class Fs implements NativeFsPromises {
  #volume: Volume;
  #promises: IPromisesAPI;
  #backend: FileSystem;

  get backend(): FileSystem {
    return this.#backend;
  }

  get volume(): Volume {
    return this.#volume;
  }

  constructor(props: FsProps = {}) {
    this.#backend = props.backend ?? 'native';
    this.#volume = new Volume({});
    this.#promises = this.#volume.promises;

    for (const method of METHODS) {
      this[method] = this.#getMethod(method);
    }
  }

  /**
   * Populate the memory file system from the provide path to the path on the
   * memory file system.
   */
  async populateFromFileSystem(props: PopulateMemory) {
    const { from, target = from, ...rest } = props;
    const promises: Array<Promise<readonly [file: string, contents: string]>> = [];
    const preserve = from === target;
    const targetPath = getPath(target);
    const iterator = glob({
      ...rest,
      cwd: from,
      trailingSlash: true,
      includeDirectories: false,
      dot: props.dot ?? true,
      caseInsensitive: props.caseInsensitive ?? true,
    });

    for await (const file of iterator) {
      const promise = fsPromises.readFile(file.absolute, 'utf8').then((contents) => {
        return [preserve ? file.absolute : path.join(targetPath, file.relative), contents] as const;
      });

      promises.push(promise);
    }

    const entries = await Promise.all(promises);
    this.#volume.fromJSON(Object.fromEntries(entries));

    return this;
  }

  /**
   * Populate the memory file system with the provided json.
   */
  populateFromJson(json: DirectoryJSON) {
    this.#volume.fromJSON(json);
    return this;
  }

  /**
   * Commit the files to the file system.
   */
  async commitToNative() {}

  /**
   * Load the current JSON stored in memory.
   */
  loadJsonFromMemory() {
    this.#volume.createReadStream;
    return this.#volume.toJSON();
  }

  /**
   * Switch to use the native node file system.
   */
  useNative() {
    this.#backend = 'native';
    return this;
  }

  useMemory() {
    this.#backend = 'memory';
    return this;
  }

  #getMethod = (method: Method): AnyFunction => {
    if (this.#backend === 'memory' && ['lutimes', 'opendir', 'watch', 'cp'].includes(method)) {
      throw new Error(`"${method}" is not supported in memory file system`);
    }

    return (...args: any[]) => {
      return this.#backend === 'native'
        ? (fsPromises as any)[method](...args)
        : (this.#promises as any)[method](...args);
    };
  };
}

interface FsProps {
  /**
   * @default 'native'
   */
  backend?: FileSystem;
}

interface PopulateMemory
  extends Pick<GlobProps, 'extensions' | 'exclude' | 'include' | 'dot' | 'caseInsensitive'> {
  /**
   * The absolute path to the file system to populate from.
   */
  from: URL | string;

  /**
   * The absolute path to target. If left undefined then this defaults to using
   * same location as from.
   */
  target?: URL | string;
}

export interface Fs extends NativeFsPromises {}
export const fs = new Fs();
export const {
  access,
  appendFile,
  chmod,
  chown,
  copyFile,
  cp,
  lchown,
  link,
  lstat,
  lutimes,
  mkdir,
  mkdtemp,
  open,
  opendir,
  readdir,
  readFile,
  readlink,
  realpath,
  rename,
  rm,
  rmdir,
  stat,
  symlink,
  truncate,
  unlink,
  utimes,
  watch,
  writeFile,
} = fs;
