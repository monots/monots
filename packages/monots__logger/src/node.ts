import { isCI, isDebug, isTest } from 'std-env';

import { Logger } from './logger.js';
import { PrettyReporter, SimpleReporter } from './node-reporters.js';
import { LogLevel } from './types.js';

const level = isDebug ? LogLevel.debug : isTest ? LogLevel.warn : LogLevel.info;

export const logger = new Logger({
  level,
  reporters: [isCI || isTest ? new SimpleReporter() : new PrettyReporter()],
});
