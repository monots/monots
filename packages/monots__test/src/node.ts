import del from 'del';
import { loadJsonFile } from 'load-json-file';
import { randomBytes } from 'node:crypto';
import path from 'node:path';
import { URL } from 'node:url';
import copy from 'recursive-copy';

interface CreateSetupFixtures<Context extends object> {
  /**
   * The current directory for the file.
   *
   * Set to `import.meta.url`.
   */
  fileUrl: string;

  /**
   * The context to use.
   */
  context: Context;

  /**
   * The name of the directory containing the fixtures.
   *
   * @default `fixtures`
   */
  fixturesDirectory?: string;
}

/**
 * Create a helper for interacting with a test fixtures directory.
 */
export function createSetupFixtures<Context extends object>(options: CreateSetupFixtures<Context>) {
  const { fileUrl, fixturesDirectory = 'fixtures', context } = options;
  let cleanups: Array<() => Promise<void>> = [];
  const root = path.join(path.dirname(new URL(fileUrl).pathname), fixturesDirectory);

  async function setupFixtures(...paths: string[]) {
    const target = path.join(root, 'tmp', randomBytes(32).toString('hex'));
    const source = path.join(root, ...paths);

    await copy(source, target);

    function getPath(...paths: string[]) {
      return path.join(target, ...paths);
    }

    async function cleanup() {
      await del(target, { force: true });
    }

    cleanups = [...cleanups, cleanup];

    return {
      getPath,
      cleanup,
      /**
       * Don't clean up the fixtures directory when using `setupFixtures.cleanup()`
       */
      skipCleanup: () => {
        cleanups = cleanups.filter((cleanup) => cleanup !== cleanup);
      },
      loadJsonFile: async <Json = any>(...paths: string[]) => loadJsonFile<Json>(getPath(...paths)),
      context: { ...context, cwd: getPath() },
    };
  }

  setupFixtures.cleanup = async () => {
    await Promise.all(cleanups.map((cleanup) => cleanup()));
  };

  return setupFixtures;
}
