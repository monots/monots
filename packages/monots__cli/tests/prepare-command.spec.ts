import { createSetupFixtures } from '@monots/test';
import * as fs from 'node:fs/promises';
import * as path from 'node:path';
import { afterAll, test } from 'vitest';

import { cli, context } from './test-setup.js';

const setupFixtures = createSetupFixtures({ context, fileUrl: import.meta.url });

afterAll(async () => {
  await setupFixtures.cleanup();
});

test.concurrent('`monots prepare` should create development dist files', async ({ expect }) => {
  const { context, getPath } = await setupFixtures('pnpm-with-packages');
  const result = await cli.run(['prepare'], context);
  const options = { encoding: 'utf8' } as const;

  const [main, types, typesWithDefault] = await Promise.all(
    [
      getPath('packages/scoped__a/dist/index.cjs.js'),
      getPath('packages/scoped__a/dist/index.d.ts'),
      getPath('packages/scoped__c/dist/other.d.ts'),
    ].map((file) => fs.readFile(file, options)),
  );

  const symlinkTargets = await Promise.all(
    [
      getPath('packages/scoped__a/dist/index.esm.js'),
      getPath('packages/scoped__a/dist/index.browser.esm.js'),
      getPath('packages/scoped__c/dist/and-another.esm.js'),
      getPath('packages/scoped__c/dist/and-another/nested.esm.js'),
      getPath('packages/scoped__c/dist/and-another/nested/cool.esm.js'),
    ].map((file) => fs.readlink(file)),
  );

  const [firstLine, ...restOfMain] = main?.split('\n') ?? [];

  expect(result, 'The result is successful').toBe(0);
  expect(firstLine?.includes('esbuild-register')).toBe(true);
  expect(firstLine?.startsWith('const { register } = require(')).toBe(true);
  expect(restOfMain.join('\n')).toMatchSnapshot();
  expect(types).toMatchSnapshot();
  expect(typesWithDefault).toMatchSnapshot();
  expect(symlinkTargets.map((file) => path.relative(getPath(), file))).toMatchSnapshot();
});
