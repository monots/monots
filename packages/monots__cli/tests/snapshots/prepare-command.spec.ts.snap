// Vitest Snapshot v1

exports[`\`monots prepare\` should create development dist files 1`] = `
"// Allow the project to be used for commonjs environments 
const register = require('../../../../../../../../node_modules/.pnpm/@swc+register@0.1.10_@swc+core@1.2.172/node_modules/@swc/register/lib/node.js');

(register.default || register)({
  \\"rootMode\\": \\"upward-optional\\",
  \\"module\\": {
    \\"type\\": \\"commonjs\\"
  },
  \\"jsc\\": {
    \\"target\\": \\"es2015\\",
    \\"transform\\": {
      \\"react\\": {
        \\"runtime\\": \\"automatic\\"
      }
    }
  }
})
module.exports = require('../src/index.ts');
register.revert();"
`;

exports[`\`monots prepare\` should create development dist files 2`] = `
"export * from \\"../src/index\\";
"
`;

exports[`\`monots prepare\` should create development dist files 3`] = `
"export * from \\"../src/other\\";
export { default } from \\"../src/other\\";
"
`;

exports[`\`monots prepare\` should create development dist files 4`] = `
[
  "packages/scoped__a/src/index.ts",
  "packages/scoped__a/src/index.ts",
  "packages/scoped__c/src/and-another/index.ts",
  "packages/scoped__c/src/and-another/nested/index.ts",
  "packages/scoped__c/src/and-another/nested/cool.ts",
]
`;
